src/Cryptol/TypeCheck/Subst.hs

```diff
@@ -435,7 +435,9 @@ instance TVars Expr where
 
         EWhere e ds   -> EWhere !$ (go e) !$ (apSubst su ds)
 
-        EPropGuards guards ty -> EPropGuards !$ (\(props, e) -> (apSubst su `fmap'` props, apSubst su e)) `fmap'` guards .$ ty
+        EPropGuards guards ty -> EPropGuards
+          !$ (\(props, e) -> (apSubst su `fmap'` props, apSubst su e)) `fmap'` guards
+          !$ apSubst su ty
```

Comment:
Perhaps slightly cleaner would be writing `go` instead of `apSubst su`, since the substitution isn't changing in any of the recursive calls:

```suggestion
          !$ (\(props, e) -> (go `fmap'` props, go e)) `fmap'` guards
          !$ go ty
```

This isn't really your fault, however, since the other cases for this function are already inconsistent about using `apSubst su` versus `go`. We should probably stick to using `go` everywhere.

Reply:
`go` has type `Expr -> Expr` so we can only use it for recursive calls on `Expr`s. The rest of the function seems to be consistent in doing that except for the call to `apSubst su e` here, so I'll just change this one.